### PR TITLE
[Profiler] Make LibrariesInfoCache standalone

### PR DESCRIPTION
**Pull request by gleocadie**
## Summary of changes

Make the cache libraries (use by libunwind for unwinding) standalone.

## Reason for change

Currently, this cache is updated before collecting a thread. It relies on the fact that all profiler goes through the same entry point to collect a thread callstack.
This was true until the new `timer-create`-based CPU profiler. This CPU profiler is triggered by the kernel (signal handler) and it's not safe updating the cache while handling its signal (avoiding memory allocation, lock taking...).
So in that situation the cache won't be updated (no update from the signal handler, no going through the same entrypoint as the other profiler) and we won't be able to collect callstacks for this profiler (except if/until another profiler calls directly to update the cache).

## Implementation details

- Add an `AutoResetEvent` class to synchronize the cache thread (its job is updating the cache) and when new `dlopen` or `dlclose` are called.
- Update the cache updates: to avoid lock ordering issue, we have to use a temporary array and swap it under the lock when it's done.
- wrap `dlopen` and `dlclose` to notify the cache when it needs to be updated.
- Use a memory pool to avoid temporary allocation

## Test coverage
- Add unit tests for the AutoResetEvent
- Add UBsan/Asan test for the cache

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (6019)](https://github.com/DataDog/dd-trace-dotnet/tree/8e87572e400effa2ef9e43c77d14e432123d7adc)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/08820b8b1d5c3d3eeafc62196c9c6853adde25d8)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6019) - mean (71ms)  : 66, 75
     .   : milestone, 71,
    master - mean (70ms)  : 66, 73
     .   : milestone, 70,

    section CallTarget+Inlining+NGEN
    This PR (6019) - mean (1,079ms)  : 1053, 1105
     .   : milestone, 1079,
    master - mean (1,080ms)  : 1050, 1111
     .   : milestone, 1080,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6019) - mean (109ms)  : 106, 112
     .   : milestone, 109,
    master - mean (109ms)  : 105, 112
     .   : milestone, 109,

    section CallTarget+Inlining+NGEN
    This PR (6019) - mean (760ms)  : 735, 785
     .   : milestone, 760,
    master - mean (753ms)  : 734, 773
     .   : milestone, 753,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6019) - mean (93ms)  : 89, 96
     .   : milestone, 93,
    master - mean (92ms)  : 90, 94
     .   : milestone, 92,

    section CallTarget+Inlining+NGEN
    This PR (6019) - mean (705ms)  : 686, 725
     .   : milestone, 705,
    master - mean (714ms)  : 693, 734
     .   : milestone, 714,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6019) - mean (190ms)  : 187, 193
     .   : milestone, 190,
    master - mean (190ms)  : 187, 194
     .   : milestone, 190,

    section CallTarget+Inlining+NGEN
    This PR (6019) - mean (1,167ms)  : 1140, 1194
     .   : milestone, 1167,
    master - mean (1,167ms)  : 1147, 1186
     .   : milestone, 1167,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6019) - mean (275ms)  : 271, 279
     .   : milestone, 275,
    master - mean (276ms)  : 272, 279
     .   : milestone, 276,

    section CallTarget+Inlining+NGEN
    This PR (6019) - mean (920ms)  : 897, 943
     .   : milestone, 920,
    master - mean (924ms)  : 902, 945
     .   : milestone, 924,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6019) - mean (264ms)  : 260, 268
     .   : milestone, 264,
    master - mean (264ms)  : 260, 268
     .   : milestone, 264,

    section CallTarget+Inlining+NGEN
    This PR (6019) - mean (903ms)  : 887, 920
     .   : milestone, 903,
    master - mean (905ms)  : 885, 925
     .   : milestone, 905,

```

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LibrariesInfoCache.h [hunk 1]
```diff
@@ -4,15 +4,20 @@
 
 #include "DlPhdrInfoWrapper.h"
 
+#include "AutoResetEvent.h"
+#include "ServiceBase.h"
+
+#include <atomic>
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LibrariesInfoCache.h:10]: remove atomic

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/AutoResetEvent.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    _impl->_isSet = true;
+    pthread_cond_signal(&_impl->_cond);
+    pthread_mutex_unlock(&_impl->_mutex);
+}
+
+bool AutoResetEvent::Wait(std::chrono::milliseconds timeout)
+{
+    pthread_mutex_lock(&_impl->_mutex);
+ 
+    bool isSignaled = true;
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/AutoResetEvent.cpp:47]: rename into isSet

File: tracer/test/snapshots/native-wrapper-symbols-arm64.verified.txt [hunk 3]
```diff
@@ -2,7 +2,6 @@ __errno_location
 access
 asprintf
 free
-getauxval
```
Comment (gleocadie) [inline tracer/test/snapshots/native-wrapper-symbols-arm64.verified.txt:5]: 🤷 do not know why it disappeared

**Review by chrisnas — APPROVED** [review]
LGTM with a few comments

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LibrariesInfoCache.cpp [hunk 4]
```diff
@@ -4,87 +4,129 @@
...
-        auto previous = NbCallsToDlopenDlclose;
-        NbCallsToDlopenDlclose = dd_nb_calls_to_dlopen_dlclose();
-        shouldReload = previous != NbCallsToDlopenDlclose;
+        // in the default case, notification mechanism in place, we will block until notification
+        // Otherwise, we reload the cache no matter on a regular basis (defaultTimeout)
+        _event.Wait(timeout);
+
+        if (_stopRequested)
+        {
+            Log::Debug("Stopping worker: stop request received.");
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LibrariesInfoCache.cpp:79]: ✔️

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LibrariesInfoCache.cpp [hunk 5]
```diff
@@ -4,87 +4,129 @@
...
 
-// This function is exposed by the ld_preload wrapper library.
-// This allows us to know if we have to reload the cache or not
-extern "C" unsigned long long dd_nb_calls_to_dlopen_dlclose() __attribute__((weak));
+extern "C" void (*volatile dd_notify_libraries_cache_update)() __attribute__((weak));
 
-LibrariesInfoCache::LibrariesInfoCache()
+LibrariesInfoCache::LibrariesInfoCache(shared::pmr::memory_resource* resource) :
+    _stopRequested{false},
+    _event(true),
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LibrariesInfoCache.cpp:19]: added a comment

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/DlPhdrInfoWrapper.cpp [hunk 6]
```diff
@@ -22,16 +22,31 @@ std::pair<struct dl_phdr_info*, std::size_t> DlPhdrInfoWrapper::Get()
...
+{
+    if (src == nullptr)
+    {
+        return {nullptr};
+    }
+
+    const std::size_t size = sizeof(T) * count;
+    auto* newPtr = static_cast<_Naked_T*>(allocator->allocate(size, alignof(T)));
+    memcpy(newPtr, src, size);
+    return {static_cast<_Naked_T*>(newPtr), [allocator, size](void* ptr) { allocator->deallocate(ptr, size, alignof(T)); }};
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/DlPhdrInfoWrapper.cpp:37]: We return a unique_ptr. So the deallocation will occur when the object will get out of scope/parent.

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/AutoResetEvent.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+#include <memory>
+
+using namespace std::chrono_literals;
+
+constexpr auto InfiniteTimeout = -1ms;
+
+class AutoResetEvent
+{
+public:
+    explicit AutoResetEvent(bool initialState);
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/AutoResetEvent.h:16]: ✔️

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/AutoResetEvent.cpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+{
+    pthread_mutex_lock(&_impl->_mutex);
+    _impl->_isSet = true;
+    pthread_cond_signal(&_impl->_cond);
+    pthread_mutex_unlock(&_impl->_mutex);
+}
+
+bool AutoResetEvent::Wait(std::chrono::milliseconds timeout)
+{
+    pthread_mutex_lock(&_impl->_mutex);
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/AutoResetEvent.cpp:46]: pthread_cond_timedwait will return immediately .

**Review by andrewlock — APPROVED** [review]
🙈

**Pull request merged**